### PR TITLE
[18.0][MIG] sell_only_by_packaging: Migration to 18.0

### PR DESCRIPTION
**Pull request by trisdoan**
### Note
- depends on: https://github.com/trisdoan/product-attribute/pull/6

### CODE REVIEW

File: sell_only_by_packaging/models/product_packaging_level.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        for record in self:
+            if record.can_be_sold:
+                continue
+            products = record.packaging_ids.product_id
+            templates = products.product_tmpl_id
+            try:
+                templates._check_sell_only_by_packaging_can_be_sold_packaging_ids()
+            except exceptions.ValidationError as e:
+                raise exceptions.ValidationError(
+                    _(
```
Comment (xaviedoanhduy) [inline sell_only_by_packaging/models/product_packaging_level.py]:
```suggestion
                    self.env._(
```

> You can replace the call to translate a string (`_`) with `self.env._` for getting some performance improvement in some cases. See https://github.com/odoo/odoo/pull/174844.

File: sell_only_by_packaging/models/product_packaging_level.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+                        'Packaging level "{}" must stay with "Can be sold",'
+                        ' at least one product configured as "sell only'
+                        ' by packaging" is using it.'
+                    ).format(record.display_name)
```
Comment (xaviedoanhduy) [inline sell_only_by_packaging/models/product_packaging_level.py]: please also get rid of the `.format`, see `https://github.com/OCA/odoo-module-migrator/pull/114`
(here and in other locations; also impact translations)

File: sell_only_by_packaging/models/product_template.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    @api.depends(
+        "sell_only_by_packaging",
+        "uom_id.factor",
+        "product_variant_ids.min_sellable_qty",
+    )
+    def _compute_template_min_sellable_qty(self):
+        for record in self:
+            record.min_sellable_qty = 0.0
+            if len(record.product_variant_ids) == 1:
```
Comment (xaviedoanhduy) [inline sell_only_by_packaging/models/product_template.py:35]:
```suggestion
            if record.product_variant_count == 1:
```

File: sell_only_by_packaging/models/product_template.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                # Pick the value from the variant if there's only 1
+                record.min_sellable_qty = record.product_variant_ids.min_sellable_qty
+
+    @api.constrains("sell_only_by_packaging", "sale_ok")
+    def _check_sell_only_by_packaging_sale_ok(self):
+        for product in self:
+            if product.sell_only_by_packaging and not product.sale_ok:
+                raise ValidationError(
+                    _(
```
Comment (xaviedoanhduy) [inline sell_only_by_packaging/models/product_template.py]:
```suggestion
                    self.env._(
```
same with all `_(` above code

File: sell_only_by_packaging/models/product_template.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    @api.constrains("sell_only_by_packaging", "packaging_ids")
+    def _check_sell_only_by_packaging_can_be_sold_packaging_ids(self):
+        for product in self:
+            if product.sell_only_by_packaging:
+                if (
+                    # Product template only condition
+                    len(product.product_variant_ids) == 1
+                    and not any(pack.sales for pack in product.packaging_ids)
+                    # Product variants condition
+                    or len(product.product_variant_ids) > 1
```
Comment (xaviedoanhduy) [inline sell_only_by_packaging/models/product_template.py:60]:
```suggestion
                    product.product_variant_count == 1
                    and not any(pack.sales for pack in product.packaging_ids)
                    # Product variants condition
                    or product.product_variant_count > 1
```

File: sell_only_by_packaging/tests/common.py
Comment (xaviedoanhduy) [inline sell_only_by_packaging/tests/common.py:1]: The code here seems to be mostly duplicated with the `product_packaging_level_salable` module, right?
https://github.com/trisdoan/product-attribute/blob/18.0-mig-product_packaging_level_salable/product_packaging_level_salable/tests/common.py

**Review by xaviedoanhduy — CHANGES_REQUESTED** [review]

File: sell_only_by_packaging/models/product_template.py (hunk 3 above)
Comment (trisdoan) [reply sell_only_by_packaging/models/product_template.py:35]: What would we gain from triggering a compute method?

**Review by xaviedoanhduy — APPROVED** [review]

**Pull request closed without merging**